From 9d4c9b504abfb76e17051f7769970a2ff7222803 Mon Sep 17 00:00:00 2001
From: OpenWrt Build System <openwrt@localhost>
Date: Thu, 30 Jul 2026 08:01:12 +0800
Subject: [PATCH 1/2] qca-nss-dp: harden EDMA v1 correctness prerequisites (F1-F6)

Split the correctness prerequisites from the RX/TX NAPI topology change. Validate DMA mappings before publishing descriptors, retain exact TX DMA addresses for unmap, decode descriptor fields with the correct endianness, reject invalid store indices, fix headroom-vs-tailroom API misuse in TX path, and unwind only IRQs that were actually requested with the matching dev_id. Keep the original single-NAPI topology and receive handoff unchanged.

Based on VIKINGYFY/immortalwrt a4638cd4389.
F1 headroom/tailroom fix from AX6 EDMA cross-audit 2026-07-26.

Signed-off-by: OpenWrt Build System <openwrt@localhost>
---
 hal/dp_ops/edma_dp/edma_v1/edma_cfg.c        | 41 +++++++-----
 hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c | 60 +++++++++++-------
 hal/dp_ops/edma_dp/edma_v1/edma_data_plane.h |  9 +--
 hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c      | 67 ++++++++++----------
 4 files changed, 100 insertions(+), 77 deletions(-)

diff --git a/hal/dp_ops/edma_dp/edma_v1/edma_cfg.c b/hal/dp_ops/edma_dp/edma_v1/edma_cfg.c
index fba67c4..62563bf 100644
--- a/hal/dp_ops/edma_dp/edma_v1/edma_cfg.c
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_cfg.c
@@ -40,6 +40,7 @@ static void edma_cleanup_rxfill_ring_res(struct edma_hw *ehw,
 	uint32_t reg_data = 0;
 	struct edma_rx_preheader *rxph = NULL;
 	int store_idx;
+	dma_addr_t dma_addr;
 
 	/*
 	 * Read RXFILL ring producer index
@@ -62,19 +63,23 @@ static void edma_cleanup_rxfill_ring_res(struct edma_hw *ehw,
 		/*
 		 * Get Rx preheader
 		 */
-		rxph = (struct edma_rx_preheader *)
-			phys_to_virt(rxfill_desc->buffer_addr);
+		dma_addr = le32_to_cpu(rxfill_desc->buffer_addr);
+		rxph = (struct edma_rx_preheader *)phys_to_virt(dma_addr);
 
-		dma_unmap_single(&pdev->dev, rxfill_desc->buffer_addr,
+		dma_unmap_single(&pdev->dev, dma_addr,
 					ehw->rx_alloc_size, DMA_FROM_DEVICE);
 
 		/*
 		 * Get sk_buff and free it
 		 */
 		store_idx = rxph->opaque;
+		if (store_idx < 0 || store_idx >= EDMA_RING_SIZE)
+			goto next_rxfill_desc;
 		skb = ehw->rx_skb_store[store_idx];
 		ehw->rx_skb_store[store_idx] = NULL;
 		dev_kfree_skb_any(skb);
+
+next_rxfill_desc:
 		cons_idx++;
 		if (cons_idx == rxfill_ring->count)
 			cons_idx = 0;
@@ -154,6 +159,7 @@ static void edma_cleanup_rxdesc_ring_res(struct edma_hw *ehw,
 	uint16_t prod_idx = 0;
 	uint16_t cons_idx = 0;
 	int store_idx;
+	dma_addr_t dma_addr;
 
 	cons_idx = edma_reg_read(EDMA_REG_RXDESC_CONS_IDX(rxdesc_ring->id))
 					& EDMA_RXDESC_CONS_IDX_MASK;
@@ -167,16 +173,19 @@ static void edma_cleanup_rxdesc_ring_res(struct edma_hw *ehw,
 	while (cons_idx != prod_idx) {
 		rxdesc_desc = EDMA_RXDESC_DESC(rxdesc_ring, cons_idx);
 
-		rxph = (struct edma_rx_preheader *)
-			phys_to_virt(rxdesc_desc->buffer_addr);
+		dma_addr = le32_to_cpu(rxdesc_desc->buffer_addr);
+		rxph = (struct edma_rx_preheader *)phys_to_virt(dma_addr);
 
-		dma_unmap_single(&pdev->dev, rxdesc_desc->buffer_addr,
+		dma_unmap_single(&pdev->dev, dma_addr,
 					ehw->rx_alloc_size, DMA_FROM_DEVICE);
 		store_idx = rxph->opaque;
+		if (store_idx < 0 || store_idx >= EDMA_RING_SIZE)
+			goto next_rxdesc_desc;
 		skb = ehw->rx_skb_store[store_idx];
 		ehw->rx_skb_store[store_idx] = NULL;
 		dev_kfree_skb_any(skb);
 
+next_rxdesc_desc:
 		/*
 		 * Update consumer index
 		 */
@@ -257,9 +266,7 @@ static void edma_cleanup_txdesc_ring_res(struct edma_hw *ehw,
 {
 	struct platform_device *pdev = ehw->pdev;
 	struct sk_buff *skb = NULL;
-	struct edma_txdesc_desc *txdesc = NULL;
 	uint16_t prod_idx, cons_idx;
-	size_t buf_len;
 	uint32_t data;
 	int store_idx;
 
@@ -273,19 +280,18 @@ static void edma_cleanup_txdesc_ring_res(struct edma_hw *ehw,
 	cons_idx = data & EDMA_TXDESC_CONS_IDX_MASK;
 
 	while (cons_idx != prod_idx) {
-		txdesc = EDMA_TXDESC_DESC(txdesc_ring, cons_idx);
-		store_idx = txdesc->buffer_addr;
+		store_idx = cons_idx;
 		skb = ehw->tx_skb_store[store_idx];
 		ehw->tx_skb_store[store_idx] = NULL;
 
-		buf_len = (txdesc->word1 & EDMA_TXDESC_DATA_LENGTH_MASK) >>
-				EDMA_TXDESC_DATA_LENGTH_SHIFT;
-
-		dma_unmap_single(&pdev->dev, (dma_addr_t)skb->data,
-				buf_len + EDMA_TX_PREHDR_SIZE, DMA_TO_DEVICE);
+		if (skb) {
+			dma_unmap_single(&pdev->dev,
+					ehw->tx_dma_store[store_idx],
+					skb_headlen(skb), DMA_TO_DEVICE);
+			ehw->tx_dma_store[store_idx] = 0;
+			dev_kfree_skb_any(skb);
+		}
 
-		dev_kfree_skb_any(skb);
-		cons_idx = (cons_idx + 1) & (txdesc_ring->count - 1);
 		cons_idx++;
 		if (cons_idx == txdesc_ring->count)
 			cons_idx = 0;
@@ -683,6 +689,7 @@ static void edma_configure_rings(struct edma_hw *ehw)
 	for (i = 0; i < EDMA_RING_SIZE; i++) {
 		ehw->tx_skb_store[i] = NULL;
 		ehw->rx_skb_store[i] = NULL;
+		ehw->tx_dma_store[i] = 0;
 	}
 
 	/*
diff --git a/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c b/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c
index 03029c9..842108f 100644
--- a/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.c
@@ -341,11 +341,9 @@ void edma_cleanup(bool is_dp_override)
 	/*
 	 * Disable Tx rings used by this driver
 	 */
-	for (i = edma_hw.txdesc_ring_start; i < edma_hw.txdesc_ring_end; i++) {
-		txcmpl_ring = &edma_hw.txcmpl_ring[i];
+	for (i = edma_hw.txdesc_ring_start; i < edma_hw.txdesc_ring_end; i++)
 		edma_reg_write(EDMA_REG_TXDESC_CTRL(i),
 				EDMA_RING_DISABLE);
-	}
 
 	/*
 	 * Disable RxFill Rings used by this driver
@@ -548,7 +548,7 @@ static netdev_tx_t edma_if_xmit(struct nss_dp_data_plane_ctx *dpc,

 	if (skb_cloned(skb) ||
 		(skb_headroom(skb) < nhead) ||
-		(skb_headroom(skb) < ntail)) {
+		(skb_tailroom(skb) < ntail)) {
 		expand_skb = true;
 	}

@@ -657,6 +655,7 @@ static int edma_irq_init(void)
 {
 	int err;
 	uint32_t entry_num, i;
+	uint32_t tx_irqs = 0, rxfill_irqs = 0, rxdesc_irqs = 0;
 
 	/*
 	 * Get TXCMPL rings IRQ numbers
@@ -671,7 +670,7 @@ static int edma_irq_init(void)
 			return -1;
 		}
 
-		pr_debug("%s: txcmpl_intr[%u] = %u\n",
+		pr_debug("%s: txcmpl_intr[%u] = %d\n",
 				 (edma_hw.device_node)->name,
 				 i, edma_hw.txcmpl_intr[i]);
 	}
@@ -688,7 +687,7 @@ static int edma_irq_init(void)
 			return -1;
 		}
 
-		pr_debug("%s: rxfill_intr[%u] = %u\n",
+		pr_debug("%s: rxfill_intr[%u] = %d\n",
 				 (edma_hw.device_node)->name,
 				 i, edma_hw.rxfill_intr[i]);
 	}
@@ -706,7 +705,7 @@ static int edma_irq_init(void)
 			return -1;
 		}
 
-		pr_debug("%s: rxdesc_intr[%u] = %u\n",
+		pr_debug("%s: rxdesc_intr[%u] = %d\n",
 				 (edma_hw.device_node)->name,
 				 i, edma_hw.rxdesc_intr[i]);
 	}
@@ -715,7 +714,12 @@ static int edma_irq_init(void)
 	 * Get misc IRQ number
 	 */
 	edma_hw.misc_intr = platform_get_irq(edma_hw.pdev, entry_num);
-	pr_debug("%s: misc IRQ:%u\n",
+	if (edma_hw.misc_intr < 0) {
+		pr_warn("%s: misc irq get failed\n",
+			(edma_hw.device_node)->name);
+		return -1;
+	}
+	pr_debug("%s: misc IRQ:%d\n",
 			  (edma_hw.device_node)->name,
 			  edma_hw.misc_intr);
 
@@ -729,9 +733,9 @@ static int edma_irq_init(void)
 		if (err) {
 			pr_debug("TXCMPL ring IRQ:%d request failed\n",
 					edma_hw.txcmpl_intr[i]);
-			return -1;
-
+			goto tx_cmpl_ring_intr_req_fail;
 		}
+		tx_irqs++;
 	}
 
 	/*
@@ -746,6 +750,7 @@ static int edma_irq_init(void)
 					edma_hw.rxfill_intr[i]);
 			goto rx_fill_ring_intr_req_fail;
 		}
+		rxfill_irqs++;
 	}
 
 	/*
@@ -760,6 +765,7 @@ static int edma_irq_init(void)
 					edma_hw.rxdesc_intr[i]);
 			goto rx_desc_ring_intr_req_fail;
 		}
+		rxdesc_irqs++;
 	}
 
 	/*
@@ -781,10 +787,11 @@ misc_intr_req_fail:
 	/*
 	 * Free IRQ for RXDESC rings
 	 */
-	for (i = 0; i < edma_hw.rxdesc_rings; i++) {
-		synchronize_irq(edma_hw.rxdesc_intr[i]);
-		free_irq(edma_hw.rxdesc_intr[i],
-				(void *)&(edma_hw.pdev)->dev);
+	while (rxdesc_irqs) {
+		rxdesc_irqs--;
+		synchronize_irq(edma_hw.rxdesc_intr[rxdesc_irqs]);
+		free_irq(edma_hw.rxdesc_intr[rxdesc_irqs],
+				(void *)edma_hw.pdev);
 	}
 
 rx_desc_ring_intr_req_fail:
@@ -792,22 +799,24 @@ rx_desc_ring_intr_req_fail:
 	/*
 	 * Free IRQ for RXFILL rings
 	 */
-	for (i = 0; i < edma_hw.rxfill_rings; i++) {
-		synchronize_irq(edma_hw.rxfill_intr[i]);
-		free_irq(edma_hw.rxfill_intr[i],
-				(void *)&(edma_hw.pdev)->dev);
+	while (rxfill_irqs) {
+		rxfill_irqs--;
+		synchronize_irq(edma_hw.rxfill_intr[rxfill_irqs]);
+		free_irq(edma_hw.rxfill_intr[rxfill_irqs],
+				(void *)edma_hw.pdev);
 	}
 
 rx_fill_ring_intr_req_fail:
+tx_cmpl_ring_intr_req_fail:
 
 	/*
 	 * Free IRQ for TXCMPL rings
 	 */
-	for (i = 0; i < edma_hw.txcmpl_rings; i++) {
-
-		synchronize_irq(edma_hw.txcmpl_intr[i]);
-		free_irq(edma_hw.txcmpl_intr[i],
-				(void *)&(edma_hw.pdev)->dev);
+	while (tx_irqs) {
+		tx_irqs--;
+		synchronize_irq(edma_hw.txcmpl_intr[tx_irqs]);
+		free_irq(edma_hw.txcmpl_intr[tx_irqs],
+				(void *)edma_hw.pdev);
 	}
 
 	return -1;
@@ -848,8 +857,11 @@ static int edma_register_netdevice(struct net_device *netdev, uint32_t macid)
 		/*
 		 * Register the interrupt handlers and enable interrupts
 		 */
-		if (edma_irq_init() < 0)
+		if (edma_irq_init() < 0) {
+			netif_napi_del(&edma_hw.napi);
+			edma_hw.netdev_arr[macid - 1] = NULL;
 			return -EINVAL;
+		}
 
 		edma_hw.napi_added = 1;
 	}
diff --git a/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.h b/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.h
index 0921e96..2266c34 100644
--- a/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.h
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_data_plane.h
@@ -205,6 +205,7 @@ struct edma_hw {
 	 */
 	struct sk_buff *rx_skb_store[EDMA_RING_SIZE];
 	struct sk_buff *tx_skb_store[EDMA_RING_SIZE];
+	dma_addr_t tx_dma_store[EDMA_RING_SIZE];
 
 	struct edma_rxfill_ring *rxfill_ring;
 			/* Rx Fill Ring, SW is producer */
@@ -239,13 +240,13 @@ struct edma_hw {
 			/* Id of first RxDesc ring */
 	uint32_t rxdesc_ring_end;
 			/* Id of last RxDesc ring */
-	uint32_t txcmpl_intr[EDMA_MAX_TXCMPL_RINGS];
+	int txcmpl_intr[EDMA_MAX_TXCMPL_RINGS];
 			/* TxCmpl ring IRQ numbers */
-	uint32_t rxfill_intr[EDMA_MAX_RXFILL_RINGS];
+	int rxfill_intr[EDMA_MAX_RXFILL_RINGS];
 			/* Rx fill ring IRQ numbers */
-	uint32_t rxdesc_intr[EDMA_MAX_RXDESC_RINGS];
+	int rxdesc_intr[EDMA_MAX_RXDESC_RINGS];
 			/* Rx desc ring IRQ numbers */
-	uint32_t misc_intr;
+	int misc_intr;
 			/* Misc IRQ number */
 
 	uint32_t tx_intr_mask;
diff --git a/hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c b/hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c
index 564e6e2..4f37f2e 100644
--- a/hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c
+++ b/hal/dp_ops/edma_dp/edma_v1/edma_tx_rx.c
@@ -40,6 +40,7 @@ int edma_alloc_rx_buffer(struct edma_hw *ehw,
 	uint32_t store_index = 0;
 	uint32_t alloc_size = ehw->rx_alloc_size;
 	struct edma_rx_preheader *rxph = NULL;
+	dma_addr_t dma_addr;
 
 	/*
 	 * Read RXFILL ring producer index
@@ -98,17 +99,14 @@ int edma_alloc_rx_buffer(struct edma_hw *ehw,
 		/*
 		 * Map Rx buffer for DMA
 		 */
-		rxfill_desc->buffer_addr = cpu_to_le32(dma_map_single(
-						&pdev->dev,
-						skb->data,
-						alloc_size,
-						DMA_FROM_DEVICE));
-
-		if (!rxfill_desc->buffer_addr) {
+		dma_addr = dma_map_single(&pdev->dev, skb->data, alloc_size,
+					  DMA_FROM_DEVICE);
+		if (dma_mapping_error(&pdev->dev, dma_addr)) {
 			dev_kfree_skb_any(skb);
 			ehw->rx_skb_store[store_index] = NULL;
 			break;
 		}
+		rxfill_desc->buffer_addr = cpu_to_le32(dma_addr);
 
 		num_alloc++;
 		next = counter;
@@ -172,9 +170,16 @@ uint32_t edma_clean_tx(struct edma_hw *ehw,
 		 * tx header contains the index in the field
 		 * buffer address (opaque) of txcmpl
 		 */
-		store_index = txcmpl->buffer_addr;
+		store_index = le32_to_cpu(txcmpl->buffer_addr);
+		if (unlikely(store_index < 0 || store_index >= EDMA_RING_SIZE)) {
+			pr_warn_ratelimited("Invalid skb index:%d cons_idx:%u prod_idx:%u\n",
+					store_index, cons_idx, prod_idx);
+			goto next_txcmpl_desc;
+		}
 		skb = ehw->tx_skb_store[store_index];
 		ehw->tx_skb_store[store_index] = NULL;
+		daddr = ehw->tx_dma_store[store_index];
+		ehw->tx_dma_store[store_index] = 0;
 
 		if (unlikely(!skb)) {
 			pr_warn_ratelimited("Invalid skb: cons_idx:%u prod_idx:%u status %x\n",
@@ -183,8 +188,6 @@ uint32_t edma_clean_tx(struct edma_hw *ehw,
 		}
 
 		len = skb_headlen(skb);
-		daddr = (dma_addr_t)virt_to_phys(skb->data);
-
 		pr_debug("skb:%px cons_idx:%d prod_idx:%d word1:0x%x\n",
 			   skb, cons_idx, prod_idx, txcmpl->status);
 
@@ -271,6 +274,7 @@ static uint32_t edma_clean_rx(struct edma_hw *ehw,
 	uint16_t cons_idx = 0;
 	uint32_t work_done = 0;
 	int store_index;
+	dma_addr_t dma_addr;
 
 	/*
 	 * Read Rx ring consumer index
@@ -297,18 +301,21 @@ static uint32_t edma_clean_rx(struct edma_hw *ehw,
 		/*
 		 * Get Rx preheader
 		 */
-		rxph = (struct edma_rx_preheader *)
-			phys_to_virt(rxdesc_desc->buffer_addr);
+		dma_addr = le32_to_cpu(rxdesc_desc->buffer_addr);
+		rxph = (struct edma_rx_preheader *)phys_to_virt(dma_addr);
 
 		/*
 		 * DMA unmap Rx buffer
 		 */
-		dma_unmap_single(&pdev->dev,
-				 rxdesc_desc->buffer_addr,
-				 ehw->rx_alloc_size,
+		dma_unmap_single(&pdev->dev, dma_addr, ehw->rx_alloc_size,
 				 DMA_FROM_DEVICE);
 
 		store_index = rxph->opaque;
+		if (unlikely(store_index < 0 || store_index >= EDMA_RING_SIZE)) {
+			pr_warn_ratelimited("Invalid Rx skb index:%d cons_idx:%u prod_idx:%u\n",
+					store_index, cons_idx, prod_idx);
+			goto next_rx_desc;
+		}
 		skb = ehw->rx_skb_store[store_index];
 		ehw->rx_skb_store[store_index] = NULL;
 		if (unlikely(!skb)) {
@@ -528,6 +535,7 @@ enum edma_tx edma_ring_xmit(struct edma_hw *ehw,
 	uint32_t data;
 	uint32_t store_index = 0;
 	struct edma_tx_preheader *txph = NULL;
+	dma_addr_t dma_addr;
 
 	/*
 	 * TODO - revisit locking
@@ -559,6 +567,12 @@ enum edma_tx edma_ring_xmit(struct edma_hw *ehw,
 		return EDMA_TX_DESC;
 	}
 
+	store_index = hw_next_to_use & (txdesc_ring->count - 1);
+	if (unlikely(ehw->tx_skb_store[store_index] != NULL)) {
+		spin_unlock_bh(&txdesc_ring->tx_lock);
+		return EDMA_TX_DESC;
+	}
+
 #if defined(NSS_DP_EDMA_TX_SMALL_PKT_WAR)
 	/*
 	 * IPQ807x EDMA hardware can't process the packet if the packet size is
@@ -600,16 +614,6 @@ enum edma_tx edma_ring_xmit(struct edma_hw *ehw,
 	txph->dst_info = (EDMA_PREHDR_DSTINFO_PORTID_IND << 8) |
 			(dp_dev->macid & 0x0fff);
 
-	/*
-	 * Store the skb in tx_store
-	 */
-	store_index = hw_next_to_use & (txdesc_ring->count - 1);
-	if (unlikely(ehw->tx_skb_store[store_index] != NULL)) {
-		spin_unlock_bh(&txdesc_ring->tx_lock);
-		return EDMA_TX_DESC;
-	}
-
-	ehw->tx_skb_store[store_index] = skb;
 	memcpy(skb->data, &store_index, 4);
 
 	/*
@@ -621,21 +625,20 @@ enum edma_tx edma_ring_xmit(struct edma_hw *ehw,
 	/*
 	 * Map buffer to DMA address
 	 */
-	txdesc->buffer_addr = cpu_to_le32(dma_map_single(&(ehw->pdev)->dev,
-				       skb->data,
-				       buf_len + EDMA_TX_PREHDR_SIZE,
-				       DMA_TO_DEVICE));
-	if (!txdesc->buffer_addr) {
+	dma_addr = dma_map_single(&(ehw->pdev)->dev, skb->data,
+				  buf_len + EDMA_TX_PREHDR_SIZE, DMA_TO_DEVICE);
+	if (dma_mapping_error(&(ehw->pdev)->dev, dma_addr)) {
 		/*
 		 * DMA map failed for this address. Drop it
 		 * and make sure does not got to stack again
 		 */
 		dev_kfree_skb_any(skb);
-
-		ehw->tx_skb_store[store_index] = NULL;
 		spin_unlock_bh(&txdesc_ring->tx_lock);
 		return EDMA_TX_OK;
 	}
+	txdesc->buffer_addr = cpu_to_le32(dma_addr);
+	ehw->tx_skb_store[store_index] = skb;
+	ehw->tx_dma_store[store_index] = dma_addr;
 
 	/*
 	 * Populate Tx descriptor
-- 
2.39.5 (Apple Git-154)

